File: src/main/java/com/pancisin/bookster/models/Service.java
package com.pancisin.bookster.models;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.pancisin.bookster.models.enums.Unit;

@Entity
@Table(name="services")
public class Service {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;
	
	@Column
	private String name;
	
	@JsonIgnore
	@ManyToOne
	private Page page;
	
	@Column
	private Double pricePerUnit;
	
	@Column
	private String detail;

	@NotNull
	@Enumerated(EnumType.STRING)
	private Unit unit;
	
	@JsonIgnore
	@OneToMany(mappedBy = "service")
	private List<BookRequest> requests;
	
	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Page getPage() {
		return page;
	}

	public void setPage(Page page) {
		this.page = page;
	}

	public Double getPricePerUnit() {
		return pricePerUnit;
	}

	public void setPricePerUnit(Double pricePerUnit) {
		this.pricePerUnit = pricePerUnit;
	}

	public String getDetail() {
		return detail;
	}

	public void setDetail(String detail) {
		this.detail = detail;
	}

	public Unit getUnit() {
		return unit;
	}

	public void setUnit(Unit unit) {
		this.unit = unit;
	}

	public List<BookRequest> getRequests() {
		return requests;
	}
}

File: src/main/kotlin/com/pancisin/bookster/model/Service.kt
package com.pancisin.bookster.model

import javax.validation.constraints.NotNull

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.pancisin.bookster.model.enums.Unit
import org.springframework.security.core.context.SecurityContextHolder
import javax.persistence.*

@Entity
@Table(name = "services")
data class Service(

  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  var id: Long? = null,

  @Column
  var name: String? = null,

  @JsonIgnore
  @ManyToOne
  var page: Page? = null,

  @Column
  var price: Double? = null,

  @Column
  var detail: String? = null,

  @NotNull
  @Enumerated(EnumType.STRING)
  var unit: Unit? = null,

  @OneToOne(cascade = [ CascadeType.ALL ], orphanRemoval = true)
  var form: Form = Form(),

  @JsonIgnore
  @OneToMany(mappedBy = "service", orphanRemoval = true, fetch = FetchType.EAGER)
  val requests: List<ServiceRequest> = ArrayList()
) {

  val requestPending : Boolean
    @JsonProperty("request_pending")
    get () {
      val auth = SecurityContextHolder.getContext().authentication.principal as User
      return requests.any { it.user?.id === auth.id && it.approved == null }
    }
}
